src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,164 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final Object[] EMPTY_ELEMENTDATA = {};
+    private static final Object[] DEFAULTCAPACITY_EMPTY_ELEMENTDATA = {};
+    private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elementData = new Object[1];
+    private int modCount = 0;
+    private int size = 0;
+
     @Override
     public void add(T value) {
-
+        modCount++;
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckForAdd(index);
+        modCount++;
+        final int s;
+        Object[] elementData;
+        if ((s = size) == (elementData = this.elementData).length) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                s - index);
+        elementData[index] = value;
+        size = s + 1;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheckForGet(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeCheckForGet(index);
+        T oldValue = (T) elementData[index];
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheckForGet(index);
+        final Object[] es = elementData;
+
+        @SuppressWarnings("unchecked") T oldValue = (T) es[index];
+        fastRemove(es, index);
+
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        final Object[] es = elementData;
+        final int size = this.size;
+        int i = 0;
+        found: {
+            if (element == null) {
+                for (; i < size; i++) {
+                    if (es[i] == null) {
+                        break found;
+                    }
+                }
+            } else {
+                for (; i < size; i++) {
+                    if (element.equals(es[i])) {
+                        break found;
+                    }
+                }
+            }
+            throw new NoSuchElementException("Element wasn't found");
+        }
+        fastRemove(es, i);
+        return element;
```

Comment:
```suggestion
        for (int i = 0; i < size; i++) {
            if (Objects.equals(elementData[i], element)) {
                return remove(i);
            }
        }
        throw new NoSuchElementException("element: '" + element + "' does not exist in the list");
```
you can simplify this method